src/deploy/functions/checkIam.ts

```diff
@@ -244,29 +267,27 @@ export async function ensureServiceAgentRoles(
     );
     return;
   }
-  // run in parallel all the missingProjectBindings jobs
-  const findRequiredBindings: Array<Promise<Array<iam.Binding>>> = [];
-  newServices.forEach((service) =>
-    findRequiredBindings.push(service.requiredProjectBindings!(projectNumber, policy))
-  );
-  const allRequiredBindings = await Promise.all(findRequiredBindings);
-  if (haveServices.length === 0) {
-    allRequiredBindings.push(obtainPubSubServiceAgentBindings(projectNumber, policy));
-    allRequiredBindings.push(obtainDefaultComputeServiceAgentBindings(projectNumber, policy));
-  }
-  if (!allRequiredBindings.find((bindings) => bindings.length > 0)) {
+  const hasUpdatedBindings = mergeBindings(policy, requiredBindings);
+  if (!hasUpdatedBindings) {
     return;
   }
-  mergeBindings(policy, allRequiredBindings);
+
   // set the updated policy
   try {
     await setIamPolicy(projectNumber, policy, "bindings");
   } catch (err: any) {
+    printManualIamConfig(requiredBindings, projectId);
     throw new FirebaseError(
       "We failed to modify the IAM policy for the project. The functions " +
         "deployment requires specific roles to be granted to service agents," +
         " otherwise the deployment will fail.",
       { original: err }
     );
   }
+
+  if (!skipSleep) {
```

Comment:
I added this in for unit testing. 

After playing around with personal accounts, I opted to remove this sleep since we aren't guaranteed to have everything setup in this timeframe. Looking deeper, there are some corner cases where we must trigger eventarc through a `create` operation before everything is generated correctly. So I enhanced our error message when we run into this case (it will only be on the first deploy) to have the user re-run the `deploy` in a few minutes (this is also what Eventarc does)